src/PhoneNumberValue.php

```diff
@@ -0,0 +1,224 @@
+        }
+
+        $isShortNumber = $short->isValidShortNumberForRegion($prototype, $regionCode)
+            || $short->isValidShortNumber($prototype);
+        $isValidNumber = $util->isValidNumberForRegion($prototype, $regionCode)
+            || $util->isValidNumber($prototype);
+
+        if (! $isShortNumber && ! $isValidNumber) {
+            throw InvalidPhoneNumberException::with(
+                $phoneNumber,
+                $regionCode
+            );
+        }
+
+        return new self(
+            $prototype,
+            $regionCode,
+            $isShortNumber
+        );
+    }
+
+    public function __toString(): string
+    {
+        if ($this->shortNumber) {
+            return $this->toNational();
+        }
+
+        return $this->toE164();
+    }
+
+    /**
+     * The ISO 3166 Country Code associated with the number
+     *
+     * @return non-empty-string
+     */
+    public function regionCode(): string
+    {
+        return $this->regionCode;
+    }
+
+    /**
+     * Whether this number is considered a 'short number' or not
+     */
+    public function isShortNumber(): bool
+    {
+        return $this->shortNumber;
+    }
+
+    public function toE164(): string
+    {
+        return PhoneNumberUtil::getInstance()->format($this->number, PhoneNumberFormat::E164);
+    }
+
+    public function toNational(): string
+    {
+        return PhoneNumberUtil::getInstance()->format($this->number, PhoneNumberFormat::NATIONAL);
+    }
+
+    public function toInternational(): string
+    {
+        return PhoneNumberUtil::getInstance()->format($this->number, PhoneNumberFormat::INTERNATIONAL);
+    }
+
+    public function toRfc3966(): string
+    {
+        return PhoneNumberUtil::getInstance()->format($this->number, PhoneNumberFormat::RFC3966);
+    }
+
+    public function toNumberDialedFrom(string $country): string
+    {
+        return PhoneNumberUtil::getInstance()->formatOutOfCountryCallingNumber($this->number, $country);
+    }
+
+    public function type(): int
+    {
+        $type = PhoneNumberUtil::getInstance()->getNumberType($this->number);
+
+        if ($this->isShortNumber()) {
+            $short = ShortNumberInfo::getInstance();
+            if ($short->isEmergencyNumber($this->toNational(), $this->regionCode)) {
+                return self::TYPE_EMERGENCY;
+            }
+
+            return self::TYPE_SHORT_CODE;
+        }
+
+        if (! array_key_exists($type, self::TYPE_MAP)) {
+            return self::TYPE_UNKNOWN;
+        }
+
+        return self::TYPE_MAP[$type];
+    }
+
+    /**
+     * @param non-empty-string|null $givenCode
+     * @return non-empty-string|null
+     */
+    private static function regionCodeForNumber(LibPhoneNumber $number, ?string $givenCode): ?string
+    {
+        $dialingCode = $number->getCountryCode();
+        if (! $dialingCode) {
+            return $givenCode;
+        }
+
```

Comment:
@gsteel hm maybe this should throw an exception if the `$regionCode !== $givenCode` because

> `+44` is not from `US` 

I like this lib and will use it when it's ready :+1:

Reply:
Thanks for reviewing @Fahl-Design!

The purpose of this method is to safely coalesce or detect the country code to a non null value if possible and is used only by `fromString($number, $maybeCountryCode)`.

The idea is:
```php
PhoneNumberValue::fromString('01234567890', 'GB'); // Valid UK, local number
PhoneNumberValue::fromString('+44 1234567890', 'CH'); // Valid UK, country argument ignored
PhoneNumberValue::fromString('01234567890', 'US'); // Exception, UK local number isn't a valid US number (I don't think)
PhoneNumberValue::fromString('01234567890', null); // Exception, local number, no idea which country
```

So returning the given code here when otherwise the code cannot be detected from the number is intentional - given code can still be null at this point.

It might be a good idea to rename the method though to better describe its purpose